File: otimization ii/linearprogramming.jl
using JuMP, GLPK

# 1. lp_data_definition

# Factory matrix: Rows = Resources (Labor, Steel), Cols = Products (Car, Truck)
factory = [
    40.0 60.0;  # Labor consumption per Car/Truck
     1.0  3.0   # Steel consumption per Car/Truck
]

# Available resources (b)
available = [1600.0, 70.0] # [Labor, Steel]

# Prices (c): Revenue per unit (to be MAXIMIZED)
prices = [13.0, 27.0]

# 2. jl_linear_programming (Using dedicated LP solver JuMP)

# Create a JuMP model using the GLPK solver
model = Model(GLPK.Optimizer)

# Define variables: car and truck, ensuring they are non-negative
@variable(model, car >= 0)
@variable(model, truck >= 0)

# Objective: Maximize Revenue (13*car + 27*truck)
@objective(model, Max, prices[1] * car + prices[2] * truck)

# Constraints (Resource Limits): factory * output <= available
# 1. Labor Constraint: 40*car + 60*truck <= 1600
@constraint(model, Labor, factory[1, 1] * car + factory[1, 2] * truck <= available[1])

# 2. Steel Constraint: 1*car + 3*truck <= 70
@constraint(model, Steel, factory[2, 1] * car + factory[2, 2] * truck <= available[2])

# Run the optimization
optimize!(model)

print("--- Julia Linear Programming Results (JuMP) ---\n")
if termination_status(model) == MOI.OPTIMAL
    print("Optimal Production Plan (car, truck):\n")
    # Extract the optimized values
    @show value(car)
    @show value(truck)
    @show objective_value(model)
else
    print("Optimization failed. Status: ", termination_status(model))
end

File: otimization ii/linearprogramming.r
# 1. lp_data_definition

# Factory matrix: Rows are resources (labor, steel), Cols are products (car, truck)
factory <- matrix(c(40,1,60,3),nrow=2,
  dimnames=list(c("labor","steel"),c("car","truck")))

# Available resources
available <- c(labor=1600, steel=70)

# Revenue per unit (to be maximized)
prices <- c(car=13,truck=27)

# Objective function: MINIMIZE the NEGATIVE revenue
revenue <- function(output) { 
  # output is a vector [car, truck]
  return(-output %*% prices) 
}

# The problem is structured as: Minimize f(output) subject to ui * output >= ci
# Constraint: factory * output <= available 
# Equivalent to: -factory * output >= -available

# 2. r_original_constrained

# Nelder-Mead optimization for constrained function
# Note: This general solver can be less efficient than a dedicated LP solver.
plan <- constrOptim(
  theta=c(5,5),      # Initial guess for [car, truck]
  f=revenue,         # Function to minimize (negative revenue)
  grad=NULL,         # Nelder-Mead does not require a gradient
  ui=-factory,       # Constraint matrix A (left side of A*x >= b)
  ci=-available,     # Constraint vector b (right side of A*x >= b)
  method="Nelder-Mead"
)

print("--- R constrOptim Results ---")
print("Optimal Production Plan (car, truck):")
# plan$par holds the optimal quantities
plan$par 